test: previous releases: add v24.0.1

theStack opened the pull request:
The same procedure as every release (see dba123167236a172d2d33861d58aa94a19729671 [v23.0] and d8b705f1caeb3b4a6790cb26e4e5584ca791d965 [v22.0]), only a little simpler now: thanks to #25650, the previous release fetch script defaults to downloading/building the necessary tags, i.e. we don't need to extend the tag list in the CI scripts and test/README.md anymore.

theStack committed:
test: previous releases: add v24.0.1

MarcoFalke commented:
I wonder if the release will get wiped, given that there is a bug in the wallet, which will be fixed in a follow-up release on the 24.x branch

Sjors commented:
Sounds good to hold off a bit. Meanwhile it needs a rebase after #26589.

luke-jr commented:
Concept NACK. Besides the bug, I'm not sure there's anything notably different between v23 and v24 to warrant adding v24 here.

Sjors commented:
@luke-jr one reason for the backwards compatibility tests is to ensure there's no accidental breakage between versions, including wallet upgrades, consensus and p2p bugs. So I think we should add every major release.

Apart from that, #25717 alone should justify adding v24.0 (or whatever ends up widely available for download).

MarcoFalke commented:
Though, `test/functional/feature_backwards_compatibility.py` doesn't test watch-only miniscript support. I don't mind, but I am not too much a fan of a "general" compat test. It is likely so general that it can't catch any issues. I much more prefer specific tests, like the ones you can find with `git grep -l skip_if_no_prev`

Sjors commented:
> It is likely so general that it can't catch any issues.

Probably not, but it's a low cost check.

> doesn't test watch-only miniscript support

I've noticed over the past few years that people would add specific tests later on. It's slightly easier to do that if the release is already there. But we can also keep this PR open until there's another PR that needs it. We've previously piled several major releases this way. Back then it was more work to rebase than it is now.

MarcoFalke commented:
> Probably not, but it's a low cost check.

Not sure if it will stay low cost forever, if the matrix is append-only with no plan to remove some or all previous releases?

Sjors commented:
> > Probably not, but it's a low cost check.
> 
> Not sure if it will stay low cost forever, if the matrix is append-only with no plan to remove some or all previous releases?

We can remove intermediate releases at some point.

luke-jr commented:
>Probably not, but it's a low cost check.

It's not low cost. To run the test at all, you have to keep a built copy of every single release listed. (And no, guix static binaries aren't a good alternative.)

Sjors commented:
Instead of hardcoding the version numbers in `feature_backwards_compatibility.py` we could have it loop over whatever is available under `releases/v*`. In that case it's up to the developer which previous releases they want to test. The default of `test/get_previous_releases.py` would still to fetch all the binaries.

MarcoFalke commented:
I don't think it is too important that a modern wallet can be opened with an ancient EOL version. It seems more important to be able to load an ancient wallet with a modern version. So maybe the EOL versions can be dropped and the qa-assets repo hosts some regtest wallets instead?

In any case the test fails intermittently (#24400), so apart from obvious smoke test issues that happen on every iteration, I don't think anyone even looks into failures here.

fanquake commented:
Noting that if this is still going ahead, it should be updated to use 24.0.1.

theStack commented:
Rebased on master and updated to v24.0.1 (note that as I'm writing this, the archives on https://bitcoincore.org/bin/ are not available yet for v24.0.1).

MarcoFalke commented on test/functional/feature_backwards_compatibility.py:
```diff
@@ -37,11 +37,12 @@ def add_options(self, parser):
         self.extra_args = [
             ["-addresstype=bech32", "-whitelist=noban@127.0.0.1"], # Pre-release: use to mine blocks. noban for immediate tx relay
             ["-nowallet", "-walletrbf=1", "-addresstype=bech32", "-whitelist=noban@127.0.0.1"], # Pre-release: use to receive coins, swap wallets, etc
+            ["-nowallet", "-walletrbf=1", "-addresstype=bech32", "-whitelist=noban@127.0.0.1"], # v24.0.1
```
I think it only makes sense to add the latest release for each missing feature. So if you want to add 24.x, you will either need to remove 23.x or explain why 24.x is special, see also 16624e6ff3af4429e571f7a606bbbcac336e067a

theStack replied:
> I think it only makes sense to add the latest release for each missing feature. So if you want to add 24.x, you will either need to remove 23.x or explain why 24.x is special, see also [16624e6](https://github.com/bitcoin/bitcoin/commit/16624e6ff3af4429e571f7a606bbbcac336e067a)

For these wallet backwards compatibility tests, wouldn't it make sense to always keep at least the recent major releases that haven't reached EOL yet as bare minimum rather than sacrificing one for another? I don't think 24.x is special to 23.x at all, but I'd rather verify this assumption (in this case e.g., that wallets created on master can still be opened on both of these versions) rather than potentially miss out on either one of those cases.

ACK on cleaning up the releases list in general, but I think that's outside of scope for this PR.

fanquake commented:
Not sure what to do here. @Sjors?

Sjors commented:
tACK 741908afc1f9ed2040c18667c75665b300c5dfe7

I checked that it still works on top of current master. That said, there's no pressing reason to merge this. As we've done before, we could just leave the PR open and occasionally rebase it.


> apart from obvious smoke test issues that happen on every iteration, I don't think anyone even looks into failures here

This can also be a reason to not rush merging new releases. This PR itself is where most of the thorough checking happens, at every rebase.

> wouldn't it make sense to always keep at least the recent major releases that haven't reached EOL yet 

I tend to agree. In any case we can always prune versions in a followup.
